### ISSUE TITLE
Location extraction: Support for big data

### ISSUE THREAD

**Issue by hong2223** [mie-lab/trackintel#61]
Currently, the function `preprocessing.staypoints.cluster_staypoints` is not able to deal with large `staypoint` input when using haversine distance, the reason is the explicit calculation and storing of the distance matrix. Using a pre-calculated distance function would also prevent the usage of 'KDTree' or 'BallTree' in the sklearn DBSCAN class. Ways to overcome the issue:
- Directly set the 'metric' parameter of the DBSCAN method to 'haversine'. This will call 'sklearn.metrics.pairwise.haversine_distances' function within DBSCAN and skip the pairwise distance matrix. But the 'epsilon' parameter needs to be converted to radius, and this conversion is likely different for each point (at different latitudes). Some libraries give a single 'epsilon' value calculated at the equator; not sure how much error will be generated。
- Recommend the user to switch to a projected coordinate system for large datasets and use euclidian distance. The error will depend on the chosen coordinate system.

**Comment by henrymartin1** [comment]
Ah, nice to see that scikits dbscan has already KNN support built in. I agree that it would be a shame not to use it.
In general I would love to see if trackintel becomes more scalable in the future which  would mean that a distance matrix calculation would be sparse by default. It could then accept either a radius or a top-k keyword by default.

- I think if it works (and provides a significant speed up) then passing the metric keyword directly to the DBSCAN algorithm is a good idea and easy to implement. If the metric is haversine and the epsilon is given in meters, then you don't have the problem of transforming epsilon, have you?
- I think for a global dataset there is nothing you can do except for taking the haversine distance,  or am I missing something?

**Comment by hong2223** [comment]
You are right that there is only one single epsilon parameter for the earth. The newest version is already submitted as a pull request.

BTW, the location clustering result using L2 (after projection) and haversine distance is the same on the SBB data. Do we know which projection we should use for the geolife data? I can then add this procedure as a test.

**Comment by hong2223** [comment]
> You are right that there is only one single epsilon parameter for the earth. The newest version is already submitted as a pull request. It's a pity that the function does not use our own distance function :/
> 
> BTW, the location clustering result using L2 (after projection) and haversine distance is the same on the SBB data. Do we know which projection we should use for the geolife data? I can then add this procedure as a test.

Fixed: We can use WGS_1984_UTM_Zone_49N (EPSG:32649) for projecting geolife

**Comment by henrymartin1** [comment]
Why do we have to project the geolife data? In general if we would assume a global dataset (Geolife is probably mostly in Beijing?), than we would not be able to project it but we would need to stay in WGS84, wouldn't we?

**Comment by hong2223** [comment]
> Why do we have to project the geolife data? In general if we would assume a global dataset (Geolife is probably mostly in Beijing?), than we would not be able to project it but we would need to stay in WGS84, wouldn't we?

It's just for a test. The location result with the projected Geolife using euclidean distance should be the same as the WGS84 Geolife using haversine distance. You can find the test in #62
